File: c/os/os_windows.go
//go:build windows

/*
 * Copyright (c) 2026 The GoPlus Authors (goplus.org). All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package os

import (
	_ "unsafe"

	"github.com/goplus/lib/c"
)

const (
	LLGoPackage = "link"
	LLGoFiles   = "_os/os_windows.c"
	PATH_MAX    = 260
)

type (
	ModeT c.Int
	UidT  c.Uint
	GidT  c.Uint
	OffT  int64
	DevT  c.Uint
)

const (
	F_GETFL = 3
	F_SETFL = 4

	O_RDONLY   = 0x0000
	O_WRONLY   = 0x0001
	O_RDWR     = 0x0002
	O_ACCMODE  = 0x0003
	O_APPEND   = 0x0008
	O_CREAT    = 0x0100
	O_TRUNC    = 0x0200
	O_EXCL     = 0x0400
	O_TEXT     = 0x4000
	O_BINARY   = 0x8000
	O_NONBLOCK = 0

	EAGAIN = 11
)

//go:linkname Clearenv C.llgo_os_clearenv
func Clearenv() c.Int

//go:linkname Getcwd C._getcwd
func Getcwd(buffer c.Pointer, size uintptr) *c.Char

//go:linkname Open C.llgo_os_open
func Open(path *c.Char, flags c.Int, __llgo_va_list ...any) c.Int

// Fcntl reports ENOSYS because the Universal CRT has no fcntl equivalent.
//
//go:linkname Fcntl C.llgo_os_fcntl
func Fcntl(fd c.Int, cmd c.Int, __llgo_va_list ...any) c.Int

//go:linkname Close C._close
func Close(fd c.Int) c.Int

//go:linkname Read C.llgo_os_read
func Read(fd c.Int, buf c.Pointer, count uintptr) int

//go:linkname Write C.llgo_os_write
func Write(fd c.Int, buf c.Pointer, count uintptr) int

//go:linkname Execlp C._execlp
func Execlp(file *c.Char, arg0 *c.Char, __llgo_va_list ...any) c.Int

File: c/os/_os/os_windows.c
/*
 * Copyright (c) 2026 The GoPlus Authors (goplus.org). All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#define _CRT_SECURE_NO_WARNINGS

#include <errno.h>
#include <fcntl.h>
#include <io.h>
#include <limits.h>
#include <stdarg.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

int llgoErrno(void) { return *_errno(); }

int llgo_os_clearenv(void)
{
    char **env = *__p__environ();
    while (env != NULL && env[0] != NULL) {
        const char *entry = env[0];
        const char *separator = strchr(entry, '=');
        if (separator == NULL || separator == entry) {
            ++env;
            continue;
        }
        size_t length = (size_t)(separator - entry);
        char *name = (char *)malloc(length + 1);
        if (name == NULL) return -1;
        memcpy(name, entry, length);
        name[length] = '\0';
        int result = _putenv_s(name, "");
        free(name);
        if (result != 0) return -1;
        env = *__p__environ();
    }
    return 0;
}

int llgo_os_open(const char *path, int flags, ...)
{
    if ((flags & _O_CREAT) != 0) {
        va_list args;
        va_start(args, flags);
        int mode = va_arg(args, int);
        va_end(args);
        return _open(path, flags, mode);
    }
    return _open(path, flags);
}

int llgo_os_fcntl(int fd, int cmd, ...)
{
    (void)fd;
    (void)cmd;
    errno = ENOSYS;
    return -1;
}

int llgo_os_read(int fd, void *buffer, size_t count)
{
    if (count > UINT_MAX) count = UINT_MAX;
    return _read(fd, buffer, (unsigned int)count);
}

int llgo_os_write(int fd, const void *buffer, size_t count)
{
    if (count > UINT_MAX) count = UINT_MAX;
    return _write(fd, buffer, (unsigned int)count);
}
